### PR TITLE
Add U-RerF

### PR DESCRIPTION
**Pull request by MrAE**
From jbrowne6's fork we have updated/added the following:
• Added U-RerF example in README
• rearranged travis.yml order to put devtools tests first
• add check --as-cran to travis.yml
• remove R/GrowUnsupervisedForest.R as it was outdated and duplicated in R/UrerfHelperFunctions.R
• remove demo/install file
• add demo/00Index file

### CODE REVIEW

File: README.Rmd [hunk 1]
```diff
@@ -24,33 +27,48 @@ knitr::opts_chunk$set(
...
 
 
 ## Tested on
 
-- Mac OSX: 10.11 (El Capitan), 10.12 (Sierra)
-- Linux: Ubuntu 16.04, CentOS 6
+- Mac OSX: 10.11 (El Capitan), 10.12 (Sierra), 10.13 (High Sierra)
+- Linux: Ubuntu 16.04 and 17.10, CentOS 6
```
Comment (falkben) [inline README.Rmd]: who tested on 17.10?  Let's take this out.

File: .#build.sh [hunk 2]
```diff
@@ -1,11 +1,12 @@
 #!/bin/bash
 
+Rscript -e "install.packages('Rcpp', repos = 'http://cran.us.r-project.org')"
```
Comment (falkben) [inline .#build.sh]: let's update ps1 file as well

File: R/UrerfHelperFunctions.R [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+  if (sizeNNZ - 1) {
+    index <- 1
+    for (m in X[1:(sizeNNZ - 1)]) {
+      leftsize <- sizeZ + index
+      rightsize <- sizeNNZ - index
+      sumLeft <- sumLeft + m
+      sumRight <- sumRight - m
+      meanLeft <- sumLeft/leftsize
+      meanRight <- sumRight/rightsize
+      errLeft <- sum((X[1:index] - meanLeft)^2) + sizeZ * (meanLeft^2)
```
Comment (falkben) [inline R/UrerfHelperFunctions.R]: Comment: error left accounts for the zeros that are missing

File: R/UrerfHelperFunctions.R [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+      errRight <- sum((X[(index + 1):sizeNNZ] - meanRight)^2)
+
+      errCurr <- errLeft + errRight
+      # Determine if this split is currently the best option
+      if (errCurr < minErr) {
+        cutPoint <- (X[index] + X[index + 1])/2
```
Comment (falkben) [inline R/UrerfHelperFunctions.R]: Comment: if current error is the lowest, we save the current cut point

File: R/UrerfHelperFunctions.R [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+
+TwoMeansCut <- function(X) {
+  minVal <- min(X)
+  maxVal <- max(X)
+  if (minVal == maxVal) {
+    return(NULL)
+  }
+  sizeX <- length(X)
+  X <- sort(X[which(X != 0)])
```
[the remark this answers is not in the document]
Comment (MrAE) [reply R/UrerfHelperFunctions.R]: decided against it.

File: R/UrerfHelperFunctions.R [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+  meanRight <- 0
+  errCurr <- 0
+  cutPoint <- NULL
+
+  if (sizeZ) {
+    meanRight <- sumRight/sizeNNZ
+    minErr <- sum((X - meanRight)^2)
+    cutPoint <- X[1]/2
+  } else {
+    minErr <- Inf
```
[the remark this answers is not in the document]
Comment (MrAE) [reply R/UrerfHelperFunctions.R]: @jbrowne6

**Comment by MrAE** [comment]
- [ ] Move `makeAB` function to the RandMat.R file and fix [[1L]] list access to make consistent.

File: R/UnsupervisedLumberjack.R [hunk 7]
```diff
@@ -0,0 +1,97 @@
+#' Unsupervised Lumberjack
```
Comment (falkben) [inline R/UnsupervisedLumberjack.R]: Delete this file and remove references in Namespace (from discussion with James this function doesn't work and shouldn't have been committed)

File: R/UrerfHelperFunctions.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrAE) [reply R/UrerfHelperFunctions.R]: If the zeros have been removed they still count for variance, making the cut point to be half of the first non-zero element seems to make sense to me.

File: R/UrerfHelperFunctions.R [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+  errRight <- 0
+  meanLeft <- 0
+  meanRight <- 0
+  errCurr <- 0
+  cutPoint <- NULL
+
+  if (sizeZ) {
+    meanRight <- sumRight/sizeNNZ
+    minErr <- sum((X - meanRight)^2)
+    cutPoint <- X[1]/2
```
[the remark this answers is not in the document]
Comment (MrAE) [reply R/UrerfHelperFunctions.R]: the zero elements still contribute to the variance.  Using half of the first non-zero element seems to make sense to me.

File: R/UrerfHelperFunctions.R [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+    OS <- FALSE
+  }
+
+  # Calculate the Max Depth and the max number of possible nodes
+  if (MaxDepth == "inf") {
+    StopNode <- 2L * w  #worst case scenario is 2*(w/(minparent/2))-1
+    MaxNumNodes <- 2L * w  # number of tree nodes for space reservation
+  } else {
+    if (MaxDepth == 0) {
+      MaxDepth <- ceiling(log2(w))
```
[the remark this answers is not in the document]
Comment (MrAE) [reply R/UrerfHelperFunctions.R]: maxDepth = 0 -> a decent stopping point for tall trees
maxDepth = Inf -> taller, but not as effective?

File: R/Urerf.R [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+#' library(rerf)
+#' urerfStructure <- Urerf(as.matrix(iris[, 1:4]))
+#'
+#' dissimilarityMatrix <-  hclust(as.dist(1-urerfStructure$similarityMatrix), method='mcquitty')
+#' clusters <- cutree(dissimilarityMatrix, k=3)
+#'
+#'
+
+Urerf <- function(X, trees = 100, min.parent = round(nrow(X)^0.5),
+                  max.depth = NA, mtry = round(ncol(X)^0.5),
```
Comment (falkben) [inline R/Urerf.R]: change to ceil for consistency

**Pull request merged**